### PR TITLE
[Kernel] Add W4A4 fused MoE mega kernel (mega_moe_w4a4)

### PR DESCRIPTION
**Pull request by Mocchibird**
# [Kernel] W4A4 fused MoE (`mega_moe_w4a4`) for Qwen3.x-MoE on Ascend

## What

Adds a single-launch fused **W4A4** (INT4 weight × INT4 activation) MoE expert kernel,
`torch.ops.npu.mega_moe_w4a4`, plus its host wrapper, op registration, and unit test. One launch
runs the entire routed-expert path — block-diagonal Hadamard, per-token INT4 quant, INT4×INT4
grouped gate-up matmul, SwiGLU + requant, INT4×INT4 down matmul, and unpermute/combine — as one
Ascend MIX kernel (cube cores do the INT4 matmuls, vector cores do everything else).

## Why

The W4A4 expert block is otherwise a chain of separate ops (init-routing, dynamic-quant,
grouped-matmul ×2, SwiGLU, token-unpermute), each a kernel launch whose activations round-trip
through global memory. Fusing the block removes the per-op host dispatch and the GM round-trips,
and keeps the INT4 intermediates on-chip. On Qwen3.6-35B-A3B (910B2, TP=4, eager) it gives
**1.14× bf16** and **1.08× the unfused per-op W4A4 chain** on GPQA-Diamond decode, for the standard
4-bit accuracy cost (2.9pp vs bf16). The 4×-smaller INT4 weights also free HBM for KV cache.

## Design

```
Stage 0 (cube)  64×64 block-diagonal Hadamard rotation of activations (fp16 GEMM)
Stage 1 (vec)   fp16 -> INT4 activations + fp32 scale, fused routing scatter
Stage 2 (cube)  INT4 x INT4 -> INT32   gate_up
Stage 3 (vec)   dequant + SwiGLU + INT4 requant
Stage 4 (cube)  INT4 x INT4 -> INT32   down
Stage 5 (vec)   dequant + scatter-add un-permute + top-k-weight combine
```

The cube matmuls are AscendC `MatmulImpl<int4b_t>`; the vector stages are PTO-ISA. The
`npu_moe_init_routing_v2` scatter is folded into Stage 1, and the per-channel down scale into the
cube FIXPIPE. The Hadamard is **calibration-free**: the rotation is baked into `gate_up` offline
and the matching rotation applied to activations online inside Stage 0 — because `(xH)(WH)^T = xW^T`
they cancel, so the kernel is self-contained (no separate Hadamard op for the graph to thread).

## Changes

New (additive only):
- `csrc/mega_moe_w4a4/op_kernel/mega_moe_w4a4_kernel.cpp` (+ `int4_cvt.hpp`) — the MIX kernel.
- `csrc/mega_moe_w4a4/op_host/mega_moe_w4a4.cpp` — `HOST_API` + shape checks + `EXEC_KERNEL_CMD`.
- `python/sgl_kernel_npu/sgl_kernel_npu/moe/mega_moe_w4a4.py` — host wrapper (tilings, workspaces,
  FRACTAL-NZ int4 weight pack, uint64 scale pack, `routing_prep`).
- `tests/python/sgl_kernel_npu/test_mega_moe_w4a4_npu.py` — cos-vs-torch unit test.

Modified (additive entries, following the `mega_chunk_gdn` pattern):
- `include/sgl_kenel_npu_ops.h` — op declaration.
- `csrc/pytorch_extensions.cpp` — `m.def` schema + `m.impl`.
- `csrc/CMakeLists.txt` — `ascendc_library(mega_moe_w4a4_kernel ...)` + link.

## Usage

```python
import torch, torch_npu  # noqa
from sgl_kernel_npu.moe.mega_moe_w4a4 import (
    routing_prep, pack_nz_int4, pack_scale_uint64, mega_moe_forward,
)
# weights (offline): w13/w2 int8 (one nibble/byte) -> FRACTAL-NZ int4; per-channel fp32 scale -> uint64
w13_nz = pack_nz_int4(w13_kn); w2_nz = pack_nz_int4(w2_kn)
w13_s  = pack_scale_uint64(w13_scale); w2_s = pack_scale_uint64(w2_scale)
# per forward: x is [T, H] fp16 (un-permuted routed activations)
group_list, expanded_row_idx, sort_idx = routing_prep(topk_ids.int(), num_experts)
y = mega_moe_forward(x, w13_nz, w13_s, w2_nz, w2_s, group_list, expanded_row_idx, sort_idx,
                     topk_weights.half(), top_k=topk_ids.shape[1])   # -> [T, H] fp16
```

## Validation

Qwen3.6-35B-A3B, Ascend 910B2, TP=4, eager, matched batch (mem 0.90, max-running 384,
context-length 16384, max-new-tokens 8192). GPQA-Diamond, 7 seeds {42,7,123,1337,1511,2024,2025};
throughput taken from the same runs as accuracy so only the MoE path differs. Both W4A4 paths read
the identical INT4 checkpoint, so accuracy is the quantization's cost, not the kernel's.

**Accuracy — standard 4-bit cost, no collapse:**

| config | GPQA-Diamond accuracy % (mean ± pstdev, 7 seeds) | Δ vs bf16 |
|---|---:|---:|
| bf16 | 70.6 ± 1.8 | — |
| W4A4 separated | 67.1 ± 1.9 | −3.5 |
| **W4A4 mega (this PR)** | **67.7 ± 2.9** | **−2.9** |

accuracy

**Throughput — decode tok/s (same runs):**

| config | decode tok/s | vs bf16 | vs W4A4 separated |
|---|---:|:--:|:--:|
| bf16 | 159 | — | — |
| W4A4 separated | 167 | 1.05× | 1.00× |
| **W4A4 mega (this PR)** | **181** | **1.14×** | **1.08×** |

throughput

The fused kernel matches the unfused W4A4 path on accuracy (67.7 vs 67.1, within seed noise — same
checkpoint) and is faster on decode. The win is modest at this batch (HBM/glue-bound) and grows at
low-batch latency and in serving capacity (4×-smaller INT4 experts free HBM for KV cache).

## Supported config / limitations

- **Atlas A2 / 910B2 (`dav-c220`) only**; build it on the target (PTO-ISA + AscendC).
- Compiled for the **Qwen3.6-35B-A3B per-rank shape: H=2048, I=128, TP=4**, Hadamard block 64. The
  wrapper asserts the shape and fails fast otherwise.
- **Routed experts only** — the shared expert / attention / LM head stay 16-bit.
- Requires the pinned **`third_party/pto-isa`** submodule (no version bump in this PR).
- KV cache is unquantized (16-bit).

## Build notes

Build-time (not JIT): `ascendc_library(mega_moe_w4a4_kernel ...)` in `csrc/CMakeLists.txt` with
`-DI_DIM_OVERRIDE=128 -DMEGA_HADAMARD_N=64 -DW4A4_MEGA_KERNEL_NAME=launch_mega_moe_w4a4`. Builds on
CANN 8.5.0 and 9.0.0 (the two CI matrix versions). Submodule pin unchanged (`third_party/pto-isa`).

## Tests

Numerically verified on Atlas A2 against a pure-torch oracle:

```bash
./build.sh -a kernels
ASCEND_RT_VISIBLE_DEVICES=0 python tests/python/sgl_kernel_npu/test_mega_moe_w4a4_npu.py
# [E=256 T=64 top_k=8] ... cos=0.999x   (production shape)  + a small-shape smoke case
```

## Running end-to-end in sglang

This PR is the kernel + host wrapper; running the full model needs sglang's NPU W4A4 MoE method to
dispatch to it. That minimal, **default-off** integration (env flag + two `NPUW4A4Int4DynamicMoEMethod`
hooks, ~80 lines) plus the full **build → link → serve → MMLU/GPQA** guide live in a draft PR to sglang:

- **Draft PR — integration code + how-to-run: https://github.com/sgl-project/sglang/pull/29045**

Enable with `SGLANG_NPU_USE_MEGA_MOE_W4A4=1` (default off → the existing torch_npu `npu_fused_experts_w4a4`
path is unchanged). The numbers above were produced through this integration.

### CODE REVIEW

**Comment by OrangeRedeng** [comment]
Hello! I am very interested in your PR as I am looking for ways to improve performance in sglang. Are you planning to create a PR in sglang? I would like to add it to our roadmap if you do not mind. Also, are you working on other types of quantization (w8a8, w4a8, wna16)?

**Comment by Mocchibird** [comment]
> Hello! I am very interested in your PR as I am looking for ways to improve performance in sglang. Are you planning to create a PR in sglang? I would like to add it to our roadmap if you do not mind. Also, are you working on other types of quantization (w8a8, w4a8, wna16)?

Hello,
I’m not currently planning to PR directly to sglang-ascend, and I’m also not working on other quantization types at the moment.
That said, I think the same general approach should be extendable to other variants, though I haven’t investigated those in detail yet...

**Comment by OrangeRedeng** [comment]
So, could I make a PR to support this kernel in sglang-ascend? Thank you for your reply!

**Comment by Mocchibird** [comment]
Sure feel free to do so.
